Armor Rendering Registry

Leo40Git opened the pull request:
A quasi-port of the [`ArmorRenderingRegistry` of old](https://github.com/FabricMC/fabric/blob/1.16/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/ArmorRenderingRegistry.java), now with LambdaFoxes Inc. brand phases!

This API resides in the new `entity_rendering` module, inside the new `rendering` library!

Leo40Git committed:
Add entity_rendering module with ArmorRenderingRegistry API

Leo40Git committed:
Begin work on testmod

Leo40Git commented:
Note that `ArmorFeatureRendererMixin` is somewhat similar to [Fabric API's `MixinArmorFeatureRenderer`](https://github.com/FabricMC/fabric/blob/1.16/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinArmorFeatureRenderer.java).
Do my changes make it unique enough, or should I add a Fabric license header to it too?

OroArmor commented:
If you copied significant amounts of code then yes

Leo40Git commented:
3 out of 5 injections are copied directly from Fabric API, but those copied injections are 2 lines each at most.
I'll add the license header, just to be safe.

Leo40Git committed:
ArmorFeatureRendererMixin: add Fabric license header

Platymemo commented:
A really nice QoL feature would be being able to define a texture path in an `ArmorMaterial`, like how vanilla uses the name but instead providing full asset path support, without having to register it.

Leo40Git commented:
I'll inject a `getTextureIdentifier()` method into `ArmorMaterial`, sure.

Leo40Git committed:
fix testmod package

make mixin abstract

Leo40Git committed:
Add getTexture method to ArmorMaterial

ArmorRenderingRegistry.TextureProvider: rename secondLayer to useSecondTexture

Leo40Git committed:
Rename internal stuff

Leo40Git committed:
Polish up testmod

SOON: TRANS ARMOR

LambdAurora commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorTextureUtils.java:
```diff
@@ -0,0 +1,33 @@
+@Environment(EnvType.CLIENT)
+public final class ArmorTextureUtils {
+	private ArmorTextureUtils() {
+		throw new RuntimeException("ArmorTextureUtils only contains static declarations.");
```
```suggestion
		throw new UnsupportedOperationException("ArmorTextureUtils only contains static declarations.");
```

LambdAurora commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorRenderingRegistry.java:
```diff
@@ -0,0 +1,84 @@
+
+	@Environment(EnvType.CLIENT)
+	public interface ModelProvider {
+		@NotNull BipedEntityModel<LivingEntity> getArmorModel(@NotNull BipedEntityModel<LivingEntity> model,
```
With the way how those are like, wrapping, maybe have

```
method (
  params
);
```

That's extremely nitpicky and subjective so if anyone wanna object feel free.

LambdAurora commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/impl/client/ArmorProviderPhaseData.java:
```diff
@@ -0,0 +1,48 @@
+
+import org.quiltmc.qsl.base.api.phase.PhaseData;
+
+public final class ArmorProviderPhaseData<T> extends PhaseData<ArmorProviderPhaseData.Provider<T>[], ArmorProviderPhaseData<T>> {
```
I admit not having expected this ahah.

I thought it would have been internal events.
We might want to switch to that actually since Events are thread-safe and this code doesn't seem to be thread-safe?
And it shouldn't be too bad to translate, unless there's a good reason to have our own handling.

Leo40Git replied:
Oh right, I left the `lock` out when copying Event's code
I couldn't think of a better way to store the applicable items.

LambdAurora replied:
Would storing an Event instance on the items work?

Leo40Git replied:
No, because `addPhaseOrdering` is independent from the applicable items.

LambdAurora replied:
but is that an issue?
since the idea of the event is you add the listener to all items separately, same for the ordering.

LambdAurora commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/impl/client/ArmorRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,36 @@
+			MODEL_PROVIDERS = new ArmorProviderManager<>();
+
+	private ArmorRenderingRegistryImpl() {
+		throw new RuntimeException("ArmorRenderingRegistryImpl only contains static declarations.");
```
```suggestion
		throw new UnsupportedOperationException("ArmorRenderingRegistryImpl only contains static declarations.");
```

LambdAurora commented on library/rendering/entity_rendering/build.gradle:
```diff
@@ -0,0 +1,20 @@
+	moduleDependencies {
+		core {
+			api("qsl_base")
+			testmodOnly("resource_loader")
```
Is resource loader as testmodOnly good? (this is a question to anyone)

Since you're most likely to use this for custom resources, having resource loader would be necessary.
Making it an api dependency would make this... I'm not sure how to word it but like, foolproof?

maximumpower55 replied:
> Is resource loader as testmodOnly good? (this is a question to anyone)
> 
> Since you're most likely to use this for custom resources, having resource loader would be necessary. Making it an api dependency would make this... I'm not sure how to word it but like, foolproof?

Yes IMO, resource loader should be a hard api dependency for this api

Leo40Git committed:
Funni is unsupported

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
how on earth did I not see this one in the files tab

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
FUNNI. IS. NOT. SUPPORTED.

Leo40Git committed:
Use events instead

ArmorRenderingRegistry.addTextureProviderPhaseOrdering and ArmorRenderingRegistry.addModelProviderPhaseOrdering are now per-item

Leo40Git committed:
ItemMixin: make getOrCreateEvent methods synchronized

Thread Safety:tm:

OroArmor commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorRenderingRegistry.java:
```diff
@@ -0,0 +1,94 @@
+		 * @param useSecondTexture {@code false} to use helmet/chestplate texture, {@code true} to use leggings/boots texture
+		 * @param suffix           suffix to append to the end of the texture name
+		 */
+		@NotNull Identifier getArmorTexture(@NotNull Identifier texture,
+				@NotNull LivingEntity entity, @NotNull ItemStack stack, @NotNull EquipmentSlot slot,
+				boolean useSecondTexture, @Nullable String suffix);
+	}
+
+	@Environment(EnvType.CLIENT)
+	public interface ModelProvider {
+		@NotNull BipedEntityModel<LivingEntity> getArmorModel(@NotNull BipedEntityModel<LivingEntity> model,
+				@NotNull LivingEntity entity, @NotNull ItemStack stack, @NotNull EquipmentSlot slot);
+	}
```
`@NotNull` is implicit. only mark `@Nullable`, unless this changed recently

LambdAurora replied:
It changed recently

for API we're getting more aggressive with annotations.
And some Kotlin users made good points about those annotations!

OroArmor commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/QuiltArmorMaterialExtensions.java:
```diff
@@ -0,0 +1,32 @@
+
+import org.quiltmc.qsl.base.api.util.InjectedInterface;
+
+@InjectedInterface(ArmorMaterial.class)
+public interface QuiltArmorMaterialExtensions {
+	@Environment(EnvType.CLIENT)
+	@NotNull Identifier getTexture();
+}
```
I would still add this interface via a mixin to armor material because injected interfaces are weird

OroArmor reviewed (approved):
Small nitpicks

Leo40Git committed:
Add default impl to ArmorMaterial.getTexture

Leo40Git committed:
Warn if armor texture (from TextureProvider) is missing file extension

Leo40Git committed:
Make Resource Loader an API dependency

Leo40Git committed:
Mark entity_rendering module as client-only

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/armor-rendering

Leo40Git committed:
Warn when ArmorMaterial.getTexture is not implemented

ArmorRenderingRegistry: rename useSecondTexture to useSecondLayer, make docs accurate
ArmorTextureUtils.getArmorTextureSuffix: don't add PNG file extension

Leo40Git committed:
I forgot applyLicenses was good now

Leo40Git committed:
javadoc spreeeeeeee

Leo40Git committed:
Merge branch '1.19' into feature/armor-rendering

Leo40Git committed:
slight change

Leo40Git committed:
Merge branch '1.19' into feature/armor-rendering

Leo40Git committed:
Merge branch '1.19' into feature/armor-rendering

Leo40Git commented:
Main API is ready, testmod needs custom texture/model which are outside of my skills...

Leo40Git commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorRenderingRegistry.java:
```diff
@@ -0,0 +1,168 @@
+	/**
+	 * The identifier of the default phase for texture/model providers.
+	 */
+	public static final Identifier DEFAULT_PHASE = Event.DEFAULT_PHASE;
```
This is a leftover from back when this API used phases directly instead of using events.
Should this be kept, though? The events themselves aren't exposed.

EnnuiL replied:
Hm, maybe not?

Leo40Git replied:
Removed!

Leo40Git commented:
...accidentally merged with 1.19.3, so changing base branch to that for now.
If this isn't good I'll redo the merge targeting 1.19 tomorrow.

But! This API shall be finished!

Leo40Git commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/QuiltArmorMaterialExtensions.java:
```diff
@@ -0,0 +1,47 @@
+	// has a "default" impl in ArmorMaterialMixin
+	// not actually marked as default, since people NEED to implement this
+	@Environment(EnvType.CLIENT)
+	@NotNull Identifier getTexture();
```
Since my last non-default injected interface methods (in Status Effect API) caused issues, should this be made default (and call `FallbackArmorTextureProvider.getArmorTexture` directly)?

Leo40Git replied:
Well I went ahead and did that anyway, so

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/armor-rendering

Leo40Git commented:
Retargeted for 1.19!

Leo40Git committed:
Environment(EnvType.CLIENT) -> ClientOnly

Satisfy checkstyle

Leo40Git committed:
Rename ItemExtensions to ItemArmorRenderingExtensions

Leo40Git committed:
Make QuiltArmorMaterialExtensions.getTexture default

Leo40Git committed:
FINALLY texture the testmod armor

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/armor-rendering

Leo40Git committed:
Texture armor *items*, too

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/armor-rendering

Leo40Git committed:
FallbackArmorTextureProvider: print material class

Leo40Git commented:
Marking as finished for now! There may be a model testmod in the future, but pretty much everything is done.

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorRenderingRegistry.java:
```diff
@@ -0,0 +1,167 @@
+	 */
+	public static void registerTextureProvider(@NotNull Identifier phaseIdentifier,
+			@NotNull TextureProvider provider, @NotNull ItemConvertible... items) {
+		for (var itemC : items) {
```
Hm, `itemC` does feel like non-standard, but heck, I don't remember what is the standard; Either way, as a small nitpick, I'd change the name to the standard name for ItemConvertibles

Leo40Git replied:
I'm unsure what the standard for this is...

EnnuiL replied:
wait heck, for ItemConvertible[], it is `items`, but for individual ones? yeah, there's nothing; I'd call it `item`, even though yeah, it's only some sort of a wrapper for the actual item itself

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorTextureUtils.java:
```diff
@@ -0,0 +1,45 @@
+	 * @return the armor texture suffix
+	 */
+	public static @NotNull String getArmorTextureSuffix(boolean useSecondLayer, @Nullable String suffix) {
+		return "_layer_" + (useSecondLayer ? 2 : 1) + (suffix == null ? "" : "_" + suffix);
```
Hm, as a precaution, I'd use this instead:
```suggestion
		return "_layer_" + (useSecondLayer ? "2" : "1") + (suffix == null ? "" : "_" + suffix);
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/impl/client/FallbackArmorTextureProvider.java:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	private static @NotNull String getArmorClassName(@NotNull ArmorMaterial material) {
+		var clazz = material.getClass();
```
oh my god it's clazz! (ignore this comment)

Leo40Git replied:
would you prefer `klass` :P

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ArmorFeatureRendererMixin.java:
```diff
@@ -0,0 +1,106 @@
+	private static Map<String, Identifier> ARMOR_TEXTURE_CACHE;
+
+	@Unique
+	private LivingEntity capturedEntity;
```
Unique variables need a `quilt$` prefix

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ArmorFeatureRendererMixin.java:
```diff
@@ -0,0 +1,106 @@
+
+	@Inject(method = "render(Lnet/minecraft/client/util/math/MatrixStack;Lnet/minecraft/client/render/VertexConsumerProvider;ILnet/minecraft/entity/LivingEntity;FFFFFF)V",
+			at = @At("HEAD"))
+	private void captureEntity(MatrixStack matrixStack, VertexConsumerProvider vertexConsumerProvider, int i, LivingEntity livingEntity, float f, float g, float h, float j, float k, float l, CallbackInfo ci) {
```
For the mixin methods, use a `quilt$` prefix

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ArmorMaterialsMixin.java:
```diff
@@ -0,0 +1,50 @@
+@Mixin(ArmorMaterials.class)
+public abstract class ArmorMaterialsMixin implements ArmorMaterial {
+	@Unique
+	private /* final */ Identifier texture;
```
Same situation here
(also, about the final, I *think* you can use a `@Final` annotation? am not sure)

Leo40Git replied:
IIRC, `@Final` can only be used with `@Shadow`.

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ItemMixin.java:
```diff
@@ -0,0 +1,65 @@
+
+@Mixin(Item.class)
+public abstract class ItemMixin implements ItemArmorRenderingExtensions {
+	@Unique
```
Ditto

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/client/ClientEntityRenderingTestmod.java:
```diff
@@ -0,0 +1,51 @@
+		ArmorRenderingRegistry.TextureProvider {
+	@Override
+	public void onInitializeClient(ModContainer mod) {
+		// TEMP DISABLED TO TEST getTexture() METHOD
```
Hm, while this does say that one will be enabled over the another, I do wonder if we can somehow use both on the armor for testing purposes

Leo40Git replied:
OH RIGHT forgot about this
I'll make the leggings texture have a nonstandard name and use the TextureProvider for that :P

EnnuiL commented on library/rendering/entity_rendering/src/testmod/resources/quilt.mod.json:
```diff
@@ -0,0 +1,28 @@
+    "intermediate_mappings": "net.fabricmc:intermediary",
+    "load_type": "always",
+    "entrypoints": {
+      "init": [
+        "org.quiltmc.qsl.rendering.entity.test.EntityRenderingTestmod"
+      ],
+      "client_init": [
+        "org.quiltmc.qsl.rendering.entity.test.client.ClientEntityRenderingTestmod"
+      ]
```
```suggestion
      "init": "org.quiltmc.qsl.rendering.entity.test.EntityRenderingTestmod",
      "client_init": "org.quiltmc.qsl.rendering.entity.test.client.ClientEntityRenderingTestmod"
```

Leo40Git replied:
Oh, I wasn't aware strings were supported here.

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/mixin/ArmorItemMixin.java:
```diff
@@ -0,0 +1,44 @@
+import net.minecraft.item.ArmorMaterial;
+import net.minecraft.item.ArmorMaterials;
+
+// TODO move this mixin to an item module
```
Succeeded by #247 :P
although i am worried about it stalling and the requested change there not happening; oh well, we really need to address those

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/QuiltArmorMaterial.java:
```diff
@@ -0,0 +1,85 @@
+/*
```
Just a complaint about Java in general: dear god there's a heckton of boilerplate

EnnuiL reviewed (changes requested):
Looks good! I only have some comments, nitpicks, and etc.

Leo40Git committed:
ArmorTextureUtils: use strings directly

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

Leo40Git committed:
Testmod QMJ: remove unnessecary arrays

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/armor-rendering

Leo40Git committed:
quilt$ prefixification

Leo40Git committed:
Reactivate test for TextureProvider

Leggings (second layer) texture uses it, because it now has a *non-standard name*

Leo40Git committed:
Rename leggings texture

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/client/ClientEntityRenderingTestmod.java:
```diff
@@ -38,7 +38,7 @@ public void onInitializeClient(ModContainer mod) {
 	}
 
-	private static final Identifier LEGGINGS_TEXTURE_ID = EntityRenderingTestmod.id("candypants.png");
+	private static final Identifier LEGGINGS_TEXTURE_ID = EntityRenderingTestmod.id("pants_made_out_of_candied_quilts.png");
```
You missed the point here; I wanted to avoid a food mention here (see: the announcement on Quiltcomm that they still have to write)

Leo40Git replied:
Whoops! Lemme fix that rq

Leo40Git committed:
Rename leggings texture, for real this time

Leo40Git committed:
ArmorRenderingRegistry: itemC -> item

Leo40Git committed:
ArmorRenderingRegsitry: remove DEFAULT_PHASE

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/client/ClientEntityRenderingTestmod.java:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2022 QuiltMC
```
```suggestion
 * Copyright 2022-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/QuiltArmorMaterial.java:
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/EntityRenderingTestmod.java:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorRenderingRegistry.java:
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2022 QuiltMC
```
```suggestion
 * Copyright 2022-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/ArmorTextureUtils.java:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2022 QuiltMC
```
```suggestion
 * Copyright 2022-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/api/client/QuiltArmorMaterialExtensions.java:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/impl/client/ArmorRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/impl/client/FallbackArmorTextureProvider.java:
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2022 QuiltMC
```
```suggestion
 * Copyright 2022-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ArmorFeatureRendererMixin.java:
```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2016, 2017, 2018, 2019 FabricMC
+ * Copyright 2022 QuiltMC
```
```suggestion
 * Copyright 2022-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/impl/client/ItemArmorRenderingExtensions.java:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ArmorMaterialsMixin.java:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ItemMixin.java:
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL commented on library/rendering/entity_rendering/src/testmod/java/org/quiltmc/qsl/rendering/entity/test/mixin/ArmorItemMixin.java:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
```suggestion
 * Copyright 2021-2023 QuiltMC
```

EnnuiL reviewed (commented):
Oh, it's 2023

Leo40Git committed:
why is it 2023

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

LambdAurora commented on library/rendering/entity_rendering/src/main/java/org/quiltmc/qsl/rendering/entity/mixin/client/ArmorFeatureRendererMixin.java:
```diff
@@ -0,0 +1,106 @@
+	@Inject(method = "getArmorTexture",
+			at = @At(value = "INVOKE", target = "Ljava/util/Map;computeIfAbsent(Ljava/lang/Object;Ljava/util/function/Function;)Ljava/lang/Object;"),
+			cancellable = true)
+	private void quilt$getArmorTexture(ArmorItem armorItem, boolean useSecondTexture, /* @Nullable */ String suffix, CallbackInfoReturnable<Identifier> cir) {
```
Why is this `@Nullable` commented?

Leo40Git replied:
Not sure! I think this method's signature is copied directly from Fabric.
I'll just remove it, this code isn't exposed to Kotlin anyway.

LambdAurora replied:
Ah that explains it then! Surely some javax stuff, uncommenting it would be nice (keepign it would be better)

Leo40Git replied:
oops

LambdAurora reviewed (approved):
Looks fine to me!

Once this is merged there's an evolution that could be made on this API: render layer swapping, [this is used in Yttr for example](https://git.sleeping.town/unascribed/Yttr/src/commit/f2d7e35909bcb21beeb4cdbfe51aa072f24331d6/src/main/java/com/unascribed/yttr/mixin/ultrapure/client/MixinArmorFeatureRenderer.java#L110). This should not be implemented in this PR since it's already ready and delaying it further isn't ideal.

Leo40Git committed:
begone commented annotation

Leo40Git committed:
return annotation

also fix a raw type

EnnuiL committed:
Merge branch '1.19' into feature/armor-rendering
